src/common/convenience.h

```diff
@@ -0,0 +1,164 @@
+// needed.
+//
+// To use this, you'll need to do something like:
+// `auto&& l = guardedly_lock(m)`
+// This way, we aren't actually copying or moving a value, we're
+// binding a reference to a temporary which extends its lifetime until
+// the end of the enclosing block.
+//
+// You may in fact want to use
+// `[[gnu::unused]] auto&& l = guardedly_lock(m)`
+// To avoid the unused variable warning. Since reference assignment
+// doesn't have side effects, normally, this just looks to the
+// compiler (whose static analysis is not the sharpest hammer in the
+// drawer) like a reference we bind for no reason and never
+// use. Perhaps future compilers will be smarter about this, but since
+// they'll also implement C++17 people might not care.
+//
+
+template<typename Mutex>
+inline auto guardedly_lock(Mutex&& m)
+  -> std::lock_guard<remove_reference_t<Mutex> > {
+  m.lock();
+  // So the way this works is that Copy List Initialization creates
+  // one and only one Temporary. There is no implicit copy that is
+  // generally optimized away the way there is if we were to just try
+  // something like `return std::lock_guard<Mutex>(m)`.
+  //
+  // The function then returns this temporary as a prvalue. We cannot
+  // bind it to a variable, because that would implicitly copy it
+  // (even if in practice RVO would mean there is no copy), so instead
+  // the user can bind it to a reference. (It has to be either a const
+  // lvalue reference or an rvalue reference.)
+  //
+  // So we get something analogous to all the others with a mildly
+  // wonky syntax. The need to use [[gnu::unused]] is honestly the
+  // worst part. It makes this construction unfortunately rather
+  // long.
+  return { std::forward<Mutex>(m), std::adopt_lock };
+}
+
+template<typename Mutex>
+inline auto guardedly_lock(Mutex&& m, std::adopt_lock_t)
+  -> std::lock_guard<remove_reference_t<Mutex> > {
+  return { std::forward<Mutex>(m), std::adopt_lock };
+}
+
+template<typename Mutex, typename Fun, typename...Args>
+inline auto with_unique_lock(Mutex&& mutex, Fun&& fun, Args&&... args)
+  -> decltype(fun(std::forward<Args>(args)...)) {
+  // Yes I know there's a lock guard inside and not a unique lock, but
+  // the caller doesn't need to know or care about the internal
+  // details, and the semantics are those of unique locking.
+  [[gnu::unused]] auto&& l = guardedly_lock(std::forward<Mutex>(mutex));
+  return std::forward<Fun>(fun)(std::forward<Args>(args)...);
+}
+
+template<typename Mutex, typename Fun, typename...Args>
+inline auto with_shared_lock(Mutex&& mutex, Fun&& fun, Args&&... args)
+  -> decltype(fun(std::forward<Args>(args)...)) {
+  auto l = sharingly_lock(std::forward<Mutex>(mutex));
+  return std::forward<Fun>(fun)(std::forward<Args>(args)...);
+}
+}
+
+// Lock Types
+// ----------
+//
+// Lock factories are nice, but you still have to type out a huge,
+// obnoxious template type when declaring a function that takes or
```

Comment:
:):) Although Its coder's Mood.
But I believe Adjectives should not be used in Comments..
This portrays the coder's Mood towards code.
Let reader read the functionality not mood of coder.